src/imitation/scripts/config/train_imitation.py

```diff
@@ -60,7 +60,11 @@ def defaults(
                 "final",
             ),
         )
-    agent_path = None
+    agent_path = None  # Path to load agent from, optional.
+
+    # Specify arguments below if agent_path for pre-trained policy is not provided.
+    policy_cls = None  # Type of a Stable Baselines3 policy architecture.
```

Comment:
We now have two `policy_cls` and `policy_kwargs`, one at the top-level and one in `train`, which is confusing:

```
$ python -m imitation.scripts.train_imitation print_config 
INFO - train_imitation - Running command 'print_config'
INFO - train_imitation - Started
Configuration (modified, added, typechanged, doc):
  agent_path = None                  # Path to load agent from, optional.
  policy_cls = None                  # Type of a Stable Baselines3 policy architecture.
  policy_kwargs = None               # Keyword arguments for policy constructor.
  [... omitted]
  train:
    n_episodes_eval = 50             # Num of episodes for final mean ground truth return
    policy_cls = <class 'imitation.policies.base.FeedForward32Policy'>    # Training
    policy_kwargs:
      features_extractor_class = <class 'imitation.policies.base.NormalizeFeaturesExtractor'>
      features_extractor_kwargs:
        normalize_class = <class 'imitation.util.networks.RunningNorm'>
INFO - train_imitation - Completed after 0:00:00
```

If you want to change the defaults for `train.{policy_cls,policy_kwargs}` you should be able to do this via a Sacred config hook: https://sacred.readthedocs.io/en/stable/apidoc.html#sacred.Experiment.config_hook

But do you actually want it to default to None...? This would mean the default (None agent_path, and None policy_cls/policy_kwargs) wouldn't run. You could provide a named_config to set them to `None`.

(Sacred doesn't make this particular use case easy, unfortunately.)

Reply:
Sorry, I did not know that policy_cls and policy_kwargs had defaults that were injected in common/train. IMO I find that the peculiar compositional nature of sacred makes it a bit hard to see everything that is going on and getting injected through decorators. I now understand your point: if you provide a policy path and the other values are automatically injected, it will always raise with my current approach. The reason I changed the default to None is to have function overloads. I still think that we should at least warn the user that the arguments are getting ignored, if enforcing it is not viable.

Reply:
Is there a way to only inject the default values that are in common/train if the path is not specified?  (and have them default to None if the path is specified). That seems like the best solution. Otherwise we can just replace the value error with a warning for now.

Reply:
> Sorry, I did not know that policy_cls and policy_kwargs had defaults that were injected in common/train. IMO I find that the peculiar compositional nature of sacred makes it a bit hard to see everything that is going on and getting injected through decorators.

Yes, the modularity has its costs. The `print_config` command is a useful way to figure out everything that's present.

> I still think that we should at least warn the user that the arguments are getting ignored, if enforcing it is not viable.

Yes agreed there should be some kind of warning and ideally an error.

> The problem is that since defaults are injected outside the function call, it is not possible to tell from within the function whether the value is a default (so ignoring is fine) or if the user is providing it (so warn/raise). It would be nice if there were a way to only inject defaults if the user does not provide a policy path. Otherwise, we can throw a warning for now whenever a path is specified, and find a better solution later.

It's possible to do this but tricky.

If we wanted to move the `agent_path` into `common/train` things would be easy. But except for BC/DAgger, we don't *want* to preload a policy -- we want to preload from an RL algorithm. So, that change wouldn't make sense.

One way to do it would be change `common/train.py` to have `policy_cls = None` be the default, then add a config hook in `common/train.py` that checks for if `agent_path` is present and non-None. This would be breaking the abstraction a bit (`common/train.py` now cares about things outside the ingredient) but would work.

You could also just change the default to be `None` and then have defaults added in the scripts as needed -- which right now is just in `common/rl.py` and `train_imitation.py`. The downside is it's harder to figure out what policy class is being used from just inspecting the config